tsl/src/fdw/data_node_scan_plan.c

```diff
@@ -741,6 +755,689 @@ push_down_group_bys(PlannerInfo *root, RelOptInfo *hyper_rel, Hyperspace *hs,
+														join_partition,
+														hyper_table_rel->serverid,
+														InvalidOid,
+														TS_FDW_RELINFO_REFJOIN_PARTITION);
+
+	Assert(join_part_fpinfo != NULL);
+
+	TsFdwRelInfo *hyper_table_rel_fpinfo = fdw_relinfo_get(hyper_table_rel);
+	Assert(hyper_table_rel_fpinfo != NULL);
+
+	/* Copy chunk assignment from hypertable */
+	join_part_fpinfo->sca = hyper_table_rel_fpinfo->sca;
+
+	/* Set parameters of the join partition */
+	join_partition->relid = hyper_table_rel->relid;
+	join_partition->relids = bms_copy(hyper_table_rel->relids);
+	join_partition->relids = bms_add_members(join_partition->relids, innerrel->relids);
+	join_partition->pathlist = NIL;
+	join_partition->partial_pathlist = NIL;
+
+	/* Set the reltarget expressions of the partition based on the reltarget expressions
+	 * of the join and adjust them for the partition */
+	join_partition->reltarget = create_empty_pathtarget();
+	join_partition->reltarget->sortgrouprefs = joinrel->reltarget->sortgrouprefs;
+	join_partition->reltarget->cost = joinrel->reltarget->cost;
+	join_partition->reltarget->width = joinrel->reltarget->width;
+#if PG14_GE
+	join_partition->reltarget->has_volatile_expr = joinrel->reltarget->has_volatile_expr;
+#endif
+	join_partition->reltarget->exprs =
+		castNode(List,
+				 adjust_appendrel_attrs(root, (Node *) joinrel->reltarget->exprs, 1, &appinfo));
+
+	return join_partition;
+}
+
+/*
+ * Create a JoinPathExtraData data structure for a partiton. The new struct is based on the original
+ * JoinPathExtraData of the join and the AppendRelInfo of the partiton.
+ */
+static JoinPathExtraData *
+create_join_partition_extra(PlannerInfo *root, JoinPathExtraData *extra, AppendRelInfo *appinfo)
+{
+	JoinPathExtraData *partition_extra = palloc(sizeof(JoinPathExtraData));
+	partition_extra->inner_unique = extra->inner_unique;
+	partition_extra->sjinfo = extra->sjinfo;
+	partition_extra->semifactors = extra->semifactors;
+	partition_extra->param_source_rels = extra->param_source_rels;
+	partition_extra->mergeclause_list =
+		castNode(List, adjust_appendrel_attrs(root, (Node *) extra->mergeclause_list, 1, &appinfo));
+	partition_extra->restrictlist =
+		castNode(List, adjust_appendrel_attrs(root, (Node *) extra->restrictlist, 1, &appinfo));
+
+	return partition_extra;
+}
+
+/*
+ * Generate the paths for a pushed down join. Each data node will be considered as a partition
+ * of the join. The join can be pushed down if:
+ *
+ * (1) The setting "ts_guc_enable_per_data_node_queries" is enabled
+ * (2) The outer relation is a distributed hypertable
+ * (3) The inner relation is marked as a reference table
+ * (4) The join is a left join or an inner join
+ *
+ * The join will be performed between the multiple DataNodeRels (see function
+ * build_data_node_part_rels) and the original innerrel of the join (the reftable).
+ */
+void
+data_node_generate_pushdown_join_paths(PlannerInfo *root, RelOptInfo *joinrel, RelOptInfo *outerrel,
```

Comment:
This is basically the top-level GetForeignJoinPaths hook, right? Can you please add the name of the corresponding fdw function and the postgres version (maybe even sha) that you used? Will be easier to port future changes this way.